### PR TITLE
Automate docker image build and push to dockerhub

### PR DESCRIPTION
**Pull request by quarkytale**
Adding Github action workflow to automatically upload mbari_wec docker image to [mbari_wec docker repository](https://hub.docker.com/repository/docker/mbari/mbari_wec/general). This would be triggered when a new tag on this repository is created.

- Need help with adding secrets on this repo for this to work. Mainly dockerhun username and password, steps [here](https://docs.github.com/en/actions/security-guides/encrypted-secrets).
- The default base is `nvidia_opengl_ubuntu22`, is that what we want?
- I haven't tested it yet, should we try the candidate release with this script? Or open to suggestions how I can test it

### CODE REVIEW

File: .github/workflows/docker_build.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: mbari/mbari_wec
+          tags: |
+            type=semver,pattern={{version}}
+            type=semver,pattern={{major}}.{{minor}}
```
Comment (mabelzhang) [inline .github/workflows/docker_build.yaml:29]: I think we said we only want to have 2 tags, `latest` and `previous`, to make maintenance easier.

For the initial test, this should probably just push to a test tag, so that it doesn't overwrite the working image we manually pushed.

**Review by mabelzhang — COMMENTED** [review]
I think we should decouple this from the first release, because this is a nice-to-have. The first release has a timeline we need to do ASAP, whereas the GitHub Actions to automatically push the image doesn't have a time constraint.

We also said we are not going to do a candidate release, because the timeline is short now. So the first release will be the official 1.0. That image on DockerHub needs to work, so we should just push that manually.

Since we already expect updates to be made to the documentation as minor releases after the 1.0, this PR could be tested with similar minor releases, e.g. `1.0.x`.

In fact, it is probably better to test this PR with dedicated prereleases, so that it isn't coupled with any actual releases. Some naming convention examples are [here on Drupal](https://www.drupal.org/docs/develop/git/git-for-drupal-project-maintainers/release-naming-conventions) and [a blog post here](https://medium.com/@nhubaovu/release-naming-conventions-2649e0a1c473). "Tags ending in -alpha, -beta, and -rc, followed by a number, are pre-releases. [Read more about the distinctions between Alpha, Beta and RC](https://www.drupal.org/node/467020)."

In short, we could test the GitHub Action with `1.0.0-rc1`, `1.0.0-rc2`, etc., increment the last `rc` number until it works. @andermi?

I don't know about the credentials... whether it is safe for us to store it here. @andermi how is the other MBARI repo storing credentials to push? Is it in a private repo? That would be a difference...

**Comment by mabelzhang** [comment]
> The default base is `nvidia_opengl_ubuntu22`, is that what we want?

Yeah, I used that for the ICRA tutorial, and that has worked for people. If GitHub Actions is able to find that image in the directory and build it successfully, then it's a tested option.

File: .github/workflows/docker_build.yaml (hunk 1 above)
Comment (quarkytale) [reply .github/workflows/docker_build.yaml]: Handled in https://github.com/osrf/mbari_wec/pull/55/commits/a701fd17ace2f8622b0d494733a101928aaa456b, the dockerhub tags would be the git tags for this repo and a `latest-rc` tag would also be created by default.

**Comment by quarkytale** [comment]
The idea is to have a `latest` tag and all previous git versions tags. Currently, a `-rc` suffix would be added by default, which should be removed after testing this workflow. Once the dockerhub login is sorted out, this can be tested.

We can also add a test job to test the image before pushing, see https://docs.docker.com/build/ci/github-actions/test-before-push/

**Comment by mabelzhang** [comment]
> all previous git versions tags

That could get too many. With the foreseeable documentation patches, the changes could be very minor. I think it could be enough to only keep the latest tag of each major version. Open to other options.

**Comment by quarkytale** [comment]
So I couldn't find how to delete or retag already uploaded images. But I can put up a check to only trigger the upload workflow on only major version changes.
